README.md
# oldsynth_patchmutator
    a utility to read Yahaha DX7 / alpha Juno patch files and mutate then using GP (Genetic Programming)


Genetic programming (GP) is a technique whereby representations of a space are encoded as a set of genes that are then modified (evolved) using an evolutionary algorithm (often a genetic algorithm, "GA")


In this case the solution space is a DX7/ a-juno synthesizer patch.

By providing "good" examples of patches and mutating/evolve/modify them new patches that are probably good will be created.


# Yamaha DX7


In order to not duplicate effort initially a compiled c++ program that is converting the dx7 binary patch  bank format to text will be used.
This was created by Ted Felix (for more info have a look here -> http://tedfelix.com/yamaha-dx7/index.html)

the output looks a bit like this:


    - Filename:romfaves.syx
    - Voice #: 1
    - Name: MOOGBASS1
    - Algorithm: 5
    - Feedback: 1
    - LFO
    - LFOWave: Square
    - LFOSpeed: 35
    - LFODelay: 0
    - LFOPitch Mod Depth: 0
    - LFOAM Depth: 0
    - LFOSync: Off
    - LFOPitch Modulation Sensitivity: 0
    - Oscillator Key Sync: On
    - Pitch Envelope Generator
    - PEGRate 1: 0
    - PEGRate 2: 0
    - PEGRate 3: 0
    - PEGRate 4: 0
    - PEGLevel 1: 50

etc...


# Roland α-juno




     0 DCO Env. Mode [0=Normal, 1=Inverted, 2=Normal-Dynamic, 3=Inverted-Dynamic]
     1 VCF Env. Mode [0=Normal, 1=Inverted, 2=Normal-Dynamic, 3=Dynamic]
     2 VCA Env. Mode [0=Normal, 1=Gate, 2=Normal-Dynamic, 3=Gate-Dynamic]
     3 DCO Wave Pulse [0..3]
     4 DCO Wave Saw   [0..5]
     5 DCO Wave Sub   [0..5]
     6 DCO Range      [0=4', 1=8', 2=16', 3=32']
     7 DCO Sub Level  [0..3]
     8 DCO Noise      [0..3]
     9 HPF Cutoff     [0..3]
    10 Chorus Switch  [0=Off, 1=On]
    11 DCO LFO Mod.   [0..127]
    12 DCO ENV Mod.   [0..127]
    13 DCO After Mod. [0..127]
    14 DCO PWM Depth  [0..127]
    15 DCO PWM Rate   [0..127] 0 = Pulse Width Manual 1..127=PW LFO Rate
    16 VCF Cutoff     [0..127]
    17 VCF Resonance  [0..127]
    18 VCF LFO Mod.   [0..127]
    19 VCF ENV Mod.   [0..127]
    20 VCF Key Follow [0..127]
    21 VCF Aftertouch [0..127]
    22 VCA Level      [0..127]
    23 VCA Aftertouch [0..127]
    24 LFO Rate       [0..127]
    25 LFO Delay      [0..127]
    26 ENV T1         [0..127] Attack Time
    27 ENV L1         [0..127] Attack Level
    28 ENV T2         [0..127] Break Time
    29 ENV L2         [0..127] Break Level
    30 ENV T3         [0..127] Decay Time
    31 ENV L3         [0..127] Sustain Level
    32 ENV T4         [0..127] Release Time
    33 ENV Key Follow [0..127]
    34 Chorus Rate    [0..127]
    35 Bender Range   [0..12]





# package dependencies / requirements

A C++ compiler (gcc / g++). To install in Ubuntu just

    sudo apt-get install gcc g++


Once patches are in textual format with some minimal parsing the rest of the work will be done with

    pandas   (data munging powerhouse)
    DEAP     (a novel evolutionary computation framework)
    Mido     (MIDI Objects for Python)


Information of the dx7 / α-juno binary format is also included in the repo for reference

There is something related on

https://github.com/rogerallen/dxsyx

I would like to use the python ecosystem however for the GA part


dx7conv/CHANGES.txt
DXconvert HISTORY
-----------------
2.1.6 (29140425)
- minor bugfixes
- started work on Korg 707/DS8 import.
  (currently only patchnames are converted)
- fixed really bad --sort bug 
- double destroy() in GUI fixed
- AudioGrill TX81Z/DX11 Voice Archive import
- Tim Thompson Glib TX81Z DX100 DX7 import
- import Grey Matter DX7 E! 32K ROM .bin file
- import Grey Matter DX7 E! RAM SysEx
- added some more .req files
- code cleanup
- improved bad patch detection

2.1.5 (20140222)
- import Atari PSSED pss480-780 library files with voicenames
- improved parameter txt print
- dx7 fixed frequency out-of-range bugfix. 
- --copy --swap commandline options: 
    copy or swap patches between locations in a bank
- CX5M .vog import (work in progress)
- Sounddiver TX81Z .lib import
- V50/YS/TQ cartridge MCD32 V50 file (.C01 ~ .C32) import
- bugfix: --yamaha=tq5|ys200|v50 bank write

2.1.4 (20131114)
- support for cassette interface data TX7, DX11, TX81Z, DX21/27/100, DX9
    wav, cas (requires wav2cas from MSX castools by Vincent van Dam)
    txt (output from tape-read perl script by M. Ward)
- minor bug fixes
- GenPatch Atari TX7 bank import
- some Grey Matter Response E! import
- --sort2 option for case-sensitive sorting (--sort is now case-insensitive)
- DX9 SysEx convert to other 4-operator synths using TXconvert
- bugfix dx7<->fourop AMD conversion
- improved --select RANGE option. RANGE can now be something like "2, 3-5, 11-32"
- fixed raw VCED import bug in dxconvert
- added README.wav2cas.txt
- fixed bug with .FB1 and .OPM uppercase output filename extension.
- --find accepts multiple keys like "string, piano"

2.1.3 (20130404)
- VOPM (*.opm) file export (TXconvert)
- improved opm import.  
- fixed txconvert-gui output file selector for DX/TX/FB01/VOPM etc.
- DX7 out-of-range level conversion management DX7/TX7/DX7II

2.1.2 (20130206)
Renata Dokmanovic says goodbye to the Yahoo YamahaDX forum.
She has helped many DX users and has had an important role during the development
of DX/TXconvert. This release is dedicated to her:
- -R or --Renata commandline option is a shortcut for --random=0
- 3rd party news: the official python-rtmidi-0.3.1a (or newer) downloads from
    http://pypi.python.org/pypi/python-rtmidi/
  should now work nicely with DX/TXconvert, also on Windows.
- MIDI in the GUI version gives mysterious problems, if run by pythonw.exe on Windows.
  They are now named dxconvert-gui.py and txconvert-gui.py again. (reported by Paul Deco)
- bugfix: unassigned variable "channel" when no channel selected in GUI.

2.1.1 (20130122)
- use extended/undocumented FB01 transpose range to compensate octave shifts after conversions
- less octave transpositions after converting from TX81Z (etc.) to DX21/27/100
- fixed a bug when rtmidi is not available. (reported and fixed by W. Cruz)

2.1.0 (201212120)
- new: Export to FB01 (TXconvert)
- improved: TX81Z > DX21 operator frequency conversion
- bugfix: FB01 > DX7 conversion OP1 level=0, not 99 
- improved: AMD parameter conversion dx7/4op/fb01
- bugfix: MIDI output in txconvert-gui.py
- improved: MIDI output reliability
- fixed deprecated --version argument
- fixed bug in 4-op parameter print
- improved: MIDI configuration dialog initialized with values from dxtxmidi.cfg,
  or with values from environment variables 'MID_IN' and 'MID_OUT'

2.0.12 (20121019)
- Some code cleanup

2.0.11 (20121011)
- Fixed really bad bug in txconvert
- Implemented Renata Dokmanovic randomizing algorithm (thanks R.!)
  (Usage: --random with value 0)

2.0.10 (20121007)
- order of excution of options changed:
    --select --find --random --nodupes --sort
  
2.0.9 (20121003)
- fix handling of banks with only INITVOICE patches.
2.0.8 (20120918)
- bump version number to 2.0.8 for official release
- add 100 ms pause between SysEx MIDI dump blocks. 
- fixed bad typo in midi_help.py
- some new .req files

2.0.7 (20120830)
2.0.7 (20120830)
- pygame.midi / pyportmidi was giving me headaches. 
  DX/TXconvert now uses python-rtmidi for the MIDI functions.
  Download it here:    
    http://pypi.python.org/pypi/python-rtmidi/
    http://chrisarndt.de/projects/python-rtmidi/download/
- Intelligent recalculation of TX81Z (and newer) operator frequencies 
  when converting for use with DX100/27/21.
- Fixed frequencies can be converted to ratios for use with DX100/27/21. 
  Approximation is based on middle E.
- Various ways to select MIDI I/O ports:
  MID_IN and MID_OUT can be set using environment variables.
  MID_IN and MID_OUT can be set using commandline options 
  MID_IN and MID_OUT can be set in the GUI version
  MID_IN and MID_OUT are saved in a file named "dxtxmidi.cfg"
  this will all override pre-defined strings in DXconvert/dxcommon.py
- fixed broken --select functions
- MIDI configuration GUI.
  
2.0.6 (20120817)
- Optimized input file format detection. 
- Randomizer default value=0. Fixes a crash that happens with Python 3
- Fixed bug in parameter list printing for 4-op synths.
- MIDI support fixed for Python3, but only partially working on Windows  

2.0.5 (20120717)
- C-Lab PolyFrame fileformat  
- Warning when too many files in a zipfile.
- MIDI I/O support (experimental, requires pygame or pyportmidi)
- fixed crash caused by some of the many available OPM files.
- fixed bad patch re-naming when converting OPM files with short file names.
- No SysEx header re-channelizing when --channel value is 0 or None.
 
2.0.4 (20120707)
- Prevent crash with bad dumps where bytes are missing in the middle of
  the dump, but SysEx header data look correct.
- minor code cleanup

2.0.3 (20120705)
- Extract and read files directly from Zip files.
- Prevent TXconvert trying to read some DX7 fileformats, 
  (for example raw .dx7 files)

2.0.2 (20120703)
- minor bugfixes and code cleanup

dx7conv/txconvert.py
#!/usr/bin/env python3
"""
txconvert.py

(c)2012 Martin Tarenskeen <m.tarenskeenATzonnetPOINTnl>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""

__license__ = 'GPL v3'

import sys
import os
from argparse import ArgumentParser
from glob import glob
from DXconvert import TXC
from DXconvert import dxcommon

MAXFILESIZE = dxcommon.MAXFILESIZE
PROGRAMNAME = TXC.PROGRAMNAME
PROGRAMVERSION = dxcommon.PROGRAMVERSION
PROGRAMDATE = dxcommon.PROGRAMDATE
ALLOWED_CHARACTERS = dxcommon.ALLOWED_CHARACTERS

############# CLI ###################

def cli_main(argv=sys.argv):
    progname = os.path.basename(argv[0])

    parser = ArgumentParser(
            description = '{}\nVersion: {} ({})'.format(PROGRAMNAME, PROGRAMVERSION, PROGRAMDATE), )
    parser.add_argument('-c', '--channel', default=0, type=int, help='Midi channel in SysEx header')
    parser.add_argument('-C', '--check', action='store_true', default=False, help='Check SysEx checksum before import') 
    parser.add_argument('-f', '--find', metavar='STRING', help='Search for STRING in patchnames')
    parser.add_argument('-mi', '--mid_in', help='select midiport MID_IN to receive data FROM synth when selecting a .req file')
    parser.add_argument('-mo', '--mid_out', help='select midiport MID_OUT to send data TO synth when choosing "MIDI" as outfile')
    parser.add_argument('-m', '--mid', help='use this option as a shortcut for "--mid_in MID_IN --mid_out MID_OUT", if MID_IN and MID_OUT have the same name MID')
    parser.add_argument('-n', '--nosplit', action='store_true', default=False, help="Don't split: save data in one file")
    parser.add_argument('-nd', '--nodupes', action='store_true', default=False, help='Remove duplicates')
    parser.add_argument('-nd2', '--nodupes2', action='store_true', default=False, help='Remove duplicates, also with different names')
    parser.add_argument('-o', '--offset', type=int, default=0, help="Ignore first OFFSET bytes in input file(s)")
    parser.add_argument('-r', '--random', metavar='NUMBER', type=int, default=-1, help="Randomize, NUMBER = a value between 0 and 300") 
    parser.add_argument('-R', '--Renata', action='store_true', default=False, help="Renata's Randomizer (same as --random=0)")
    parser.add_argument('-sp', '--split', action='store_true', default=False, help="Save each patch in a separate file")
    parser.add_argument('-s', '--select', metavar='RANGE', help='Select patch RANGE to save')
    parser.add_argument('-S', '--sort', action='store_true', default=False, help='Sort patches by name, not case-sensitive')
    parser.add_argument('-S2', '--sort2', action='store_true', default=False, help='Sort patches by name, case-sensitive')
    parser.add_argument('--copy', nargs=2, metavar=('X', 'Y'), help='Copy patch nr X to nr Y')
    parser.add_argument('--swap', nargs=2, metavar=('X', 'Y'), help='Swap patch nr X and nr Y')
    parser.add_argument('-t', '--tx7', action='store_true', default=False, help='Save with TX7 Performance data')
    parser.add_argument('-v', '--version', action='version', version='{} ({})'.format(PROGRAMVERSION, PROGRAMDATE)) 
    parser.add_argument('-y', '--yamaha', metavar='MODEL', default='tx81z', help='Specify Yamaha synth MODEL for outfile') 
    parser.add_argument('infile', nargs='+', help='Selects input file(s)')
    parser.add_argument('outfile', help='Select output file')

    args=parser.parse_args()
    infilez = args.infile
    infiles = []
    for i in infilez:
        infiles += glob(i)
    outfile = args.outfile

    mid_in = dxcommon.MID_IN
    mid_out = dxcommon.MID_OUT
    CFG = dxcommon.CFG
    if os.getenv('MID_IN'):
        mid_in = os.getenv('MID_IN')
    mid_out = os.getenv('MID_OUT')
    if os.path.exists(CFG):
        with open(CFG, 'r') as f:
            for line in f.readlines():
                l = line.split('=')
                if l[0].strip() == 'MID_IN':
                    mid_in = l[1].strip()
                if l[0].strip() == 'MID_OUT':
                    mid_out = l[1].strip()

    if args.mid:
        mid_in = args.mid
        mid_out =  args.mid
    if args.mid_in:
        mid_in = args.mid_in
    if args.mid_out:
        mid_out = args.mid_out
    if args.mid_in or args.mid_out:
        with open(CFG, 'w') as f:
            if mid_in: f.write('MID_IN = {}\n'.format(mid_in))
            if mid_out: f.write('MID_OUT = {}\n'.format(mid_out))

    nosplit = args.nosplit
    ch = args.channel
    check = args.check
    select = args.select
    sort = args.sort
    sort2 = args.sort2
    swap = args.swap
    copy = args.copy
    nodupes = args.nodupes
    nodupes2 = args.nodupes2
    offset = args.offset
    deviation = max(-1, min(300, args.random))
    if args.Renata:
        deviation = 0
    find = args.find
    yamaha = args.yamaha.lower().strip()
    FB01 = False
    if args.split:
        split = 1
    elif yamaha in ('ys100', 'ys200', 'tq5', 'b200', 'ds55', 'v50'):
        split = 25
    elif yamaha in ('dx27s', 'dx27', 'dx27s', 'dx100'):
        split = 24
    elif yamaha == 'fb01' or os.path.splitext(outfile)[1].lower() == '.fb1':
        yamaha = 'fb01'
        FB01 = True
        split = 48
    elif yamaha == 'vopm' or os.path.splitext(outfile)[1].lower() == '.opm':
        yamaha = 'vopm'
        FB01 = True
        split = 128
    else:
        split = 32

    txdata = []
    for infile in infiles:
        print("Reading {}".format(infile))
        if not os.path.exists(infile):
            print("{} not found".format(infile))
            return 1
        if os.path.getsize(infile)>MAXFILESIZE:
            if MAXFILESIZE != 0:
                print("Warning: File too large.\nOnly {} bytes will be read".format(MAXFILESIZE))
        if infile == outfile:
            print("Must have different input and output files")
            return 1

        if os.path.isfile(infile):
            txdat, channel = TXC.read(infile, offset, check, yamaha, mid_in, mid_out)
            txdata += txdat


    if ch != None:
        channel = ch-1
        channel = max(0, channel)
        channel = min(15, channel)

    if nodupes2:
        nodupes = True
            
    if nodupes or nodupes2:
        if FB01:
            txdata = TXC.fbnodupes(txdata, nodupes2)
        else:
            txdata = TXC.txnodupes(txdata, nodupes2)

    if select != None: 
        txdat = []
        for i in dxcommon.range2list(select):
            if FB01:
                txdat += txdata[64*(i-1): 64*i]
            else:
                txdat += txdata[128*(i-1): 128*i]
        txdata = txdat
    
    if args.find != None:
        if FB01:
            txdata = TXC.fbfind(txdata, find)
        else:
            txdata = TXC.txfind(txdata, find)
    
    if deviation != -1:
        if FB01:
            txdata = TXC.fbrandom(txdata, deviation)
        else:
            txdata = TXC.txrandom(txdata, deviation)

    if sort:
        if FB01:
            txdata = TXC.fbsort(txdata, False)
        else:
            txdata = TXC.txsort(txdata, False)
    if sort2:
        if FB01:
            txdata = TXC.fbsort(txdata, True)
    if copy:
        if FB01:
            txdata = TXC.fbcopy(dxcommon.range2list(copy[0]), int(copy[1]), txdata)
        else:
            txdata = TXC.txcopy(dxcommon.range2list(copy[0]), int(copy[1]), txdata)

    if swap:
        if FB01:
            txdata = TXC.fbswap(int(swap[0]), int(swap[1]), txdata)
        else:
            txdata = TXC.txswap(int(swap[0]), int(swap[1]), txdata)

    if split == 1:
        if FB01:
            for i in range(len(txdata)//64):
                outfile_name = dxcommon.list2string(txdata[64*i:64*i+7])
                for j in range(len(outfile_name)):
                    if not outfile_name[j] in ALLOWED_CHARACTERS:
                        outfile_name = outfile_name[:j] + "_" + outfile_name[j+1:]
                count = 0
                Outfile_Name = outfile_name
                while os.path.exists(Outfile_Name+outfile):
                    count += 1
                    if count>1: Outfile_Name = outfile_name + "(" + str(count) + ")"

                outfile_name = os.path.join(os.path.split(outfile)[0], Outfile_Name + os.path.split(outfile)[1])

                TXC.write(outfile_name, txdata[64*i:64*(i+1)], channel, nosplit, 1, yamaha, mid_out)
 
        else:
            for i in range(len(txdata)//128):
                outfile_name = dxcommon.list2string(txdata[128*i+57:128*i+67])
                for j in range(len(outfile_name)):
                    if not outfile_name[j] in ALLOWED_CHARACTERS:
                        outfile_name = outfile_name[:j] + "_" + outfile_name[j+1:]
                count = 0
                Outfile_Name = outfile_name
                while os.path.exists(Outfile_Name+outfile):
                    count += 1
                    if count>1: Outfile_Name = outfile_name + "(" + str(count) + ")"

                outfile_name = os.path.join(os.path.split(outfile)[0], Outfile_Name + os.path.split(outfile)[1])

                TXC.write(outfile_name, txdata[128*i:128*(i+1)], channel, nosplit, 1, yamaha, mid_out)
    else:
        TXC.write(outfile, txdata, channel, nosplit, split, yamaha, mid_out)

    return 0


################### MAIN #####################

if __name__ == '__main__':
    sys.exit(cli_main())
    

